src/alloc.rs

```diff
@@ -1,87 +1,93 @@
-use crate::{base::Base, internal, RawMem, Result};
+use crate::{Error::CapacityOverflow, RawMem, Result};
 use std::{
     alloc::{Allocator, Layout},
     cmp::Ordering,
+    marker::PhantomData,
+    mem,
+    mem::MaybeUninit,
     ptr::{drop_in_place, NonNull},
 };
 use tap::Pipe;
 
 pub struct Alloc<T, A: Allocator> {
-    base: Base<T>,
+    ptr: NonNull<T>,
+    len: usize,
     alloc: A,
+    _marker: PhantomData<T>,
 }
 
 impl<T: Default, A: Allocator> Alloc<T, A> {
-    pub const fn new(alloc: A) -> Self {
+    pub const fn new(len: usize, alloc: A) -> Self {
         Self {
-            base: Base::dangling(),
+            ptr: NonNull::dangling(),
+            len,
             alloc,
+            _marker: PhantomData,
         }
     }
 
-    unsafe fn alloc_impl(&mut self, capacity: usize) -> Result<&mut [T]> {
-        let old_capacity = self.base.ptr.len();
-        let new_capacity = capacity;
-
-        let result: Result<_> = try {
-            if self.base.ptr.as_non_null_ptr() == NonNull::dangling() {
-                let layout = Layout::array::<T>(capacity)?;
-                self.alloc.allocate(layout)?
-            } else {
-                let old_layout = Layout::array::<T>(old_capacity)?;
-                let new_layout = Layout::array::<T>(new_capacity)?;
-
-                let ptr = internal::to_bytes(self.base.ptr);
-                match new_capacity.cmp(&old_capacity) {
-                    Ordering::Less => {
-                        self.base.handle_narrow(new_capacity);
-                        self.alloc
-                            .shrink(ptr.as_non_null_ptr(), old_layout, new_layout)?
-                    }
-                    Ordering::Greater => {
-                        self.alloc
-                            .grow(ptr.as_non_null_ptr(), old_layout, new_layout)?
-                    }
-                    Ordering::Equal => ptr,
-                }
+    fn current_memory(&self) -> Option<(NonNull<u8>, Layout)> {
+        if self.len == 0 {
+            None
+        } else {
+            unsafe {
```

Comment:
Add safety message about this `unsafe` block  
```suggestion
            // SAFETY: we would use `Layout::array`, but memory is allocated yet
            // and it's size+align is always valid (because we already alloc it by `Layout::array`)
            unsafe {
```